criu/log: properly handle truncated length from vsnprintf

avagin opened the pull request:
vsnprintf does not always return the number of bytes actually written to the buffer.

If the output was truncated due to the buffer limit, the return value is the total number of bytes which WOULD have been written to the final string if enough space had been available.

This means we must cap the return value to the buffer size excluding the terminating null byte to correctly calculate the log entry size.

Fixes #2781

avagin committed:
criu/log: properly handle truncated length from vsnprintf

vsnprintf does not always return the number of bytes actually written to
the buffer.

If the output was truncated due to the buffer limit, the return value is
the total number of bytes which WOULD have been written to the final
string if enough space had been available.

This means we must cap the return value to the buffer size excluding the
terminating null byte to correctly calculate the log entry size.

Signed-off-by: Andrei Vagin <avagin@google.com>

rst0git reviewed (approved):
LGTM


I was able to confirm that this change fixes the error in https://github.com/checkpoint-restore/criu/issues/2781

kolyshkin commented on criu/log.c:
```diff
@@ -320,7 +320,7 @@ unsigned int log_get_loglevel(void)
+	int log_size = 0, log_space;
 	struct early_log_hdr *hdr;
 
 	if ((early_log_buf_off + sizeof(hdr)) >= EARLY_LOG_BUF_LEN)
```
It looks like usage of `sizeof(hdr)` is also wrong here (and in other two places), because `hdr` is a pointer (8 bytes on x86_64). It should be replaced with `sizeof(struct early_log_hdr)` (4 bytes) like in your new code above.

kolyshkin replied:
...or `sizeof(*hdr)`

avagin replied:
They are consistent in all places and they use the size that is greater than the right one, so everything works as it should be :)

kolyshkin reviewed (commented):
One more (minor) issue found.

avagin committed:
log: use sizeof(*hdr) instead of sizeof(hdr)

Using sizeof(hdr) where hdr is a pointer gives the size of the pointer,
not the size of the structure it points to.

Reported-by: Kir Kolyshkin <kolyshkin@gmail.com>
Signed-off-by: Andrei Vagin <avagin@gmail.com>
